### ISSUE TITLE
Troubleshooting Reproducibility Issues

### ISSUE THREAD

**Issue by k0rtal** [CPS-TUWien/racing_dreamer#13]
I have kept the hyperparameters and the fundamental code unchanged, but I am still unable to achieve the expected reproduction. 

Could you kindly provide any guidance on how to properly reproduce the results?

One notable issue I encountered is that the provided Dockerfile was not fully compatible with my computer(RTX 4080 SUPER). As a result, I made the following modifications(requirements.txt is not changed):

``` dockerfile
FROM nvcr.io/nvidia/tensorflow:20.11-tf2-py3

RUN python3 --version && pip3 --version && pip3 list
RUN apt update -y && apt install git wget unzip -y

RUN python3 -m pip install pip==21.3.1

# install requirements
WORKDIR /build
COPY requirements.txt /build
RUN pip3 install -r requirements.txt

WORKDIR /build/src/racecar-gym/models/scenes
RUN wget https://github.com/axelbr/racecar_gym/releases/download/tracks-v1.0.0/all.zip && unzip all.zip

# Switch to src directory
WORKDIR /src

# add working directory to the python path
ENV PYTHONPATH "${PYTHONPATH}:/src"
ENV LD_LIBRARY_PATH=/usr/local/cuda/lib64:/usr/local/nvidia/lib:/usr/local/nvidia/lib64:$LD_LIBRARY_PATH

# Copy your code into the docker that is assumed to live in . (on machine)
COPY ./ /src
```

I would greatly appreciate any insights or advice you could share. 

Thank you for your time and assistance!

**Comment by axelbr** [comment]
Could you please specify what's not working (building or running), the command you run, the error you are encountering?

**Comment by axelbr** [comment]
Please also have a look at this issue first: https://github.com/CPS-TUWien/racing_dreamer/issues/12

**Comment by k0rtal** [comment]
https://github.com/user-attachments/assets/3f15dd4c-9279-4973-bf2b-5568787efc86

![Image](https://github.com/user-attachments/assets/021ae176-24ee-41e9-a09f-7c075d61e477)

![Image](https://github.com/user-attachments/assets/74b76a26-50fc-48f1-9506-1009b561f1df)

## **It looks like learning is not going well.**

I trained 5,000,000 epochs.


One notable issue I encountered is that the provided Dockerfile was not fully compatible with my computer(RTX 4080 SUPER). So I changed dockerfile as above.

The following are the outputs that appear when running the terminal.

``` bash
kim@kim-SSSSS-SP-PPPPP:~/racing_dreamer/dreamer$ docker run -u $(id -u):$(id -g) -v $(pwd):/src --env="DISPLAY" --gpus all --volume="/tmp/.X11-unix:/tmp/.X11-unix:rw" --shm-size=1g --ulimit memlock=-1 --ulimit stack=67108864 --rm dreamer:v0 python3 dream.py --track columbia --obs_type lidar --steps 5000000
                                                                                                                                                
================
== TensorFlow ==
================

NVIDIA Release 20.11-tf2 (build 17379986)
TensorFlow Version 2.3.1

Container image Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
Copyright 2017-2020 The TensorFlow Authors.  All rights reserved.

Various files include modifications (c) NVIDIA CORPORATION.  All rights reserved.
NVIDIA modifications are covered by the license terms that apply to the underlying project or file.
WARNING: Detected NVIDIA NVIDIA GeForce RTX 4080 SUPER GPU, which is not yet supported in this version of the container
ERROR: No supported GPU(s) detected to run this container

NOTE: MOFED driver for multi-node communication was not detected.
      Multi-node communication performance may be reduced.

2025-02-05 09:47:58.371411: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
Matplotlib created a temporary config/cache directory at /tmp/matplotlib-was1ui6f because the default path (/.config/matplotlib) is not a writable directory; it is highly recommended to set the MPLCONFIGDIR environment variable to a writable directory, in particular to speed up the import of Matplotlib and to better support multiprocessing.
pybullet build time: Nov 28 2023 23:49:17
2025-02-05 09:48:04.569263: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcuda.so.1
2025-02-05 09:48:04.592023: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1079] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2025-02-05 09:48:04.592122: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1742] Found device 0 with properties: 
pciBusID: 0000:01:00.0 name: NVIDIA GeForce RTX 4080 SUPER computeCapability: 8.9
coreClock: 2.55GHz coreCount: 80 deviceMemorySize: 15.69GiB deviceMemoryBandwidth: 685.51GiB/s
2025-02-05 09:48:04.592134: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
2025-02-05 09:48:04.611161: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcublas.so.11
2025-02-05 09:48:04.624736: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcufft.so.10
2025-02-05 09:48:04.628905: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcurand.so.10
2025-02-05 09:48:04.650497: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcusolver.so.11
2025-02-05 09:48:04.657065: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcusparse.so.11
2025-02-05 09:48:04.657393: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudnn.so.8
2025-02-05 09:48:04.657578: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1079] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2025-02-05 09:48:04.657889: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1079] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2025-02-05 09:48:04.658044: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1884] Adding visible gpu devices: 0
2025-02-05 09:48:04.700624: I tensorflow/core/platform/profile_utils/cpu_utils.cc:104] CPU Frequency: 1996800000 Hz
2025-02-05 09:48:04.703353: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x2da448e0 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2025-02-05 09:48:04.703393: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
2025-02-05 09:48:04.745787: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1079] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2025-02-05 09:48:04.745906: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x2df6da10 initialized for platform CUDA (this does not guarantee that XLA will be used). Devices:
2025-02-05 09:48:04.745916: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): NVIDIA GeForce RTX 4080 SUPER, Compute Capability 8.9
2025-02-05 09:48:04.746521: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1079] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2025-02-05 09:48:04.746573: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1742] Found device 0 with properties: 
pciBusID: 0000:01:00.0 name: NVIDIA GeForce RTX 4080 SUPER computeCapability: 8.9
coreClock: 2.55GHz coreCount: 80 deviceMemorySize: 15.69GiB deviceMemoryBandwidth: 685.51GiB/s
2025-02-05 09:48:04.746584: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
2025-02-05 09:48:04.746594: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcublas.so.11
2025-02-05 09:48:04.746599: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcufft.so.10
2025-02-05 09:48:04.746603: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcurand.so.10
2025-02-05 09:48:04.746607: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcusolver.so.11
2025-02-05 09:48:04.746611: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcusparse.so.11
2025-02-05 09:48:04.746616: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudnn.so.8
2025-02-05 09:48:04.746635: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1079] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2025-02-05 09:48:04.746679: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1079] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2025-02-05 09:48:04.746711: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1884] Adding visible gpu devices: 0
2025-02-05 09:48:04.747196: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
2025-02-05 09:48:06.495955: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1283] Device interconnect StreamExecutor with strength 1 edge matrix:
2025-02-05 09:48:06.496011: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1289]      0 
2025-02-05 09:48:06.496022: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1302] 0:   N 
2025-02-05 09:48:06.497351: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1079] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2025-02-05 09:48:06.497668: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1079] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2025-02-05 09:48:06.497869: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1428] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 13609 MB memory) -> physical GPU (device: 0, name: NVIDIA GeForce RTX 4080 SUPER, pci bus id: 0000:01:00.0, compute capability: 8.9)
2025-02-05 09:48:17.174905: W tensorflow/python/util/util.cc:348] Sets are not currently considered sequences, but this may change in the future, so consider avoiding using them.
2025-02-05 09:48:30.532464: W tensorflow/core/grappler/optimizers/loop_optimizer.cc:906] Skipping loop optimization for Merge node with control input: cond_6/then/_4868/cond_6/cond/branch_executed/_15772
2025-02-05 09:48:33.210928: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcublas.so.11
2025-02-05 09:49:01.654926: W tensorflow/core/grappler/optimizers/loop_optimizer.cc:906] Skipping loop optimization for Merge node with control input: cond_6/then/_4868/cond_6/cond/branch_executed/_15772
[Info] Logdir dreamer/logs/experiments/columbia_dreamer_max_progress_lidar_tanhnormal_Ar4_Bl50_H15_825704_1738748884.6584463
[Info] Prefill dataset (strategy=gap_follower) with 5000 steps.
	[Summary] Train episode of length 500 (2000 sim steps) with return 58.7.
	[Summary] Train episode of length 500 (2000 sim steps) with return 64.0.
	[Summary] Train episode of length 500 (2000 sim steps) with return 63.0.
[Init] Found 2533922 model parameters.
[Init] Found 575204 actor parameters.
[Init] Found 413601 value parameters.
[Info] Simulating agent for 994000 steps.
[Info] Start evaluation.
	[Summary] Test episode of length 57 (228 sim steps) with return 6.1.
	[Summary] Test episode of length 44 (176 sim steps) with return 3.7.
	[Summary] Test episode of length 51 (204 sim steps) with return 4.7.
	[Summary] Test episode of length 47 (188 sim steps) with return 4.2.
	[Summary] Test episode of length 43 (172 sim steps) with return 3.1.
[Info] Found New Best Model: 4.67810
[Info] Start collection.
[Info] Training for 100 steps.
	[Train Step] # 0
GIF summaries require ffmpeg in $PATH. [Errno 2] No such file or directory: 'ffmpeg': 'ffmpeg'
GIF summaries require ffmpeg in $PATH. [Errno 2] No such file or directory: 'ffmpeg': 'ffmpeg'
	[Train Step] # 1
	[Train Step] # 2
	[Train Step] # 3
	[Train Step] # 4
	[Train Step] # 5

...etc...

```

This link is my events.out.tfevents file
https://drive.google.com/file/d/1wQ90Zz7TiZx0mYU_JxBNBYh-4gw1ZvM0/view?usp=sharing

**Comment by axelbr** [comment]
> One notable issue I encountered is that the provided Dockerfile was not fully compatible with my computer(RTX 4080 SUPER). So I changed dockerfile as above.

This does not look like it's impacting training much.

Regarding the performance:
The tensorboard logs indicate that it achieves a mean max-progress of 1.3 within the evaluation horizon during training, which corresponds to 1.3 laps. It looks like it performs worse in the evaluation mode. 
You could try to visualize a training episode.

Unfortunately we are working on different projects now and do not have the capacity to provide support except fixing simple bugs/dependency issues.

**Comment by k0rtal** [comment]
Thank you for advice!

If I fix it, I will pull request.
